src/main/java/org/truffleruby/core/cast/DurationToMillisecondsNode.java

```diff
@@ -65,12 +67,25 @@ protected long durationNil(Nil duration) {
     }
 
     @Specialization
-    protected long duration(RubyDynamicObject duration) {
+    protected long duration(RubyDynamicObject duration,
+            @Cached DispatchNode divmodNode,
+            @Cached DispatchNode sumNode,
+            @Cached ToLongNode toLongNode) {
         if (floatCastNode == null) {
             CompilerDirectives.transferToInterpreterAndInvalidate();
             floatCastNode = insert(NumericToFloatNodeGen.create());
         }
-        return duration(floatCastNode.executeDouble(duration));
+        try {
+            return duration(floatCastNode.executeDouble(duration));
+        } catch (Exception e) {
```

Comment:
Catching `Exception` is dangerous, e.g., it might hide a NullPointerException.
Catching `RaiseException` wouldn't be great either as it could catch other Ruby exceptions.

What we should do here is to follow the logic in CRuby (what you wrote below) and don't try the NumericToFloatNode/floatCastNode at all.
BTW, this seems the only usage of NumericToFloatNode, so we can just remove that class.